### PR TITLE
relayCall events overhaul

### PR DESCRIPTION
**Pull request by nventuro**
Currently, `TransactionRelayed`'s meaning depends on whether the `status` field is equal to a specific value (`CanRelayFailed`) or not. The `chargeOrCanRelayStatus` field, with dual meaning that requires looking at the _other_  status code in order to be read is a consequence of this. 

This PR splits `TransactionRelayed` (emitted by `relayCall`) into two distinct events:

* `RelayCallFailed`: emitted when the relayed call was not executed and the recipient not charged. The reason is included. Equivalent to the old `TransactionRelayed` with `status` set to `CanRelayFailed`
* `TransactionRelayed`: emitted when the call was relayed, and the recipient charged. The execution status and charge are included. Equivalent to the old `TransactionRelayed` with `status` set to something other than `CanRelayFailed`

### CODE REVIEW

**Comment by drortirosh** [comment]
We discussed internally this issue (single or multiple events for success/failure)
The intent behind having a single event, was to allow a client to have a simple query to check for reputation: I can query for all success and failure events of a specific (or any) relay, and decide based on that how worthwhile it is to use it.
Now such a client will have to double the # of API calls to query the relays status.
The same events can be used by a Relay to detect malicious contract/client pair.
For relay (with persistent storage), its much easier to collect such data, so double the calls is meaningless, but for clients it might be important.

While technically possible, client APIs can't have "OR" filter on the event, only on the (indexed) parameters

**Comment by frangio** [comment]
>client APIs can't have "OR" filter on the event, only on the (indexed) parameters

@drortirosh Are you sure about this? The RPC allows filtering on log topics, and for Solidity events these include the event identifier as well as the indexed event parameters. So at the level of RPC it should be possible to "OR filter" the event. It may not be possible to do with the web3.js contract abstraction, but I don't think that should hold us back.

**Comment by drortirosh** [comment]
@frangio, look at the API if getPastEvents: first parameter is "topic", then a set of options.
While the rpc doesn't care and doesn't differentiate the event topic from indexed parameters, the web3 does - it uses it to decode the parameters and event name

**Comment by spalladino** [comment]
I wouldn't base the design of the contract on the capabilities of a medium-level library, especially since the library can be easily extended to support that use case (which I expect it to be, since it's a feature supported by the underlying JSON-RPC layer), and the contract is much more long-lived.

Also, even though web3 does not support filtering by multiple event names on the contract object, it is possible to [rely on `getPastLogs`](https://web3js.readthedocs.io/en/1.0/web3-eth.html#getpastlogs) and build the filter manually from the contract ABI.

File: contracts/IRelayHub.sol [hunk 1]
```diff
@@ -113,17 +113,21 @@ contract IRelayHub {
     // Emits a TransactionRelayed event.
     function relayCall(address from, address to, bytes memory encodedFunction, uint256 transactionFee, uint256 gasPrice, uint256 gasLimit, uint256 nonce, bytes memory approval) public;
 
+    // Emitted when an attempt to relay a call failed. This can happen due to incorrect relayCall arguments, or the
+    // recipient not accepting the relayed call. The actual relayed call was not executed, and the recipient not charged.
+    // The reason field contains an error code: values 1-10 correspond to PreconditionCheck entries, and values over 10
+    // are custom recipient error codes returned from acceptRelayedCall.
+    event RelayCallFailed(address indexed relay, address indexed from, address indexed to, bytes4 selector, uint256 reason);
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/IRelayHub.sol]: >There are other failures that are not reported by this event.

Do you mean the other checks (recipient has enough balance, gas price is correct, available gas is enough)? I have a separate branch where those are also included in the event, as opposed to being exposed via a revert reason.

>clients: must check the event after transmitting a transaction, to check whether it was reverted or not (that is transaction succeeded only if it got TransactionRelayed AND also status==0. But in order to report an error - it must get the RelayedCallFailed event for its reason field - or the above TransactionRelayed.status field...

These are two very different scenarios though: in one case the relay sent a malformed transaction to `relayCall` and it was rejected (might be a relay attempting to cause delays), so it needs to go look for a new relay, while in the other there is an issue with the data the client signed and sent to the relay. Handling of those should not be the same.

File: contracts/RelayHub.sol [hunk 2]
```diff
@@ -303,7 +303,7 @@ contract RelayHub is IRelayHub {
         balances[recipient] -= charge;
         balances[relays[msg.sender].owner] += charge;
 
-        emit TransactionRelayed(msg.sender, from, recipient, functionSelector, uint256(status), charge);
+        emit TransactionRelayed(msg.sender, from, recipient, functionSelector, status, charge);
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/RelayHub.sol]: That depends on what you mean by 'success': it is not the relay's responsibility to make sure the relayed tx doesn't revert, nor is it the client's to make sure the precondition checks/canRelay succeed.

In which case would someone look for failures, defined by either `RelayCallFailed` being emitted or `TransactionRelayed`'s status not being zero?

File: contracts/IRelayHub.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (drortirosh) [reply contracts/IRelayHub.sol]: I only meant about the statuses reported by these 2 events. there are "static" checks, like insufficient funds that are not reported here, and can easily checked by the client in advance.

The problem we raise here is that we can't simply have "failed" event and "success" event: that's OK for a client, but a relay actually want to distinguish between "GotPaid" and "DidntGetPaid" transactions - a relayed call that reverted (either the call, pre or post) is considered "success" for a relay - but not for a client!

So we either have 3 different events - or get back to one.
the 3 types are `CanRelayFailed`, `TransactionSuccess`, `TransactionFailedButPaid` - but I don't think we should go that way..

File: contracts/RelayHub.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (drortirosh) [reply contracts/RelayHub.sol]: We are creating an events API for 2 usages: clients and relays.
So we wanted to make the events "unbiased" for these use-cases. We don't want to use Success or Failure in the event name if a use-case doesn't treat it as such.
> In which case would someone look for failures, defined by either RelayCallFailed being emitted or TransactionRelayed's status not being zero?
Yes, that's what a client does: in the old API, only events with `status==0` are considered success.
for a relay, `status!=CanRelayFailed` means success ... (where success in this case means "I got paid, no-one tried to screw with me.")

**Comment by nventuro** [comment]
(replying here since we're having the same discussion in two places in parallel)

I think an underlying issue is that there are multiple error scenarios that require specific action from each party. I'll try to summarize the concerns of both relays and clients (using the events proposed here), assuming honest and bug-free implementations. This same criteria can also be applied to past events to evaluate whether or not they trust each other.

(Also, note that the current RelayHub reverts under certain conditions, such as insufficient recipient balance. I will assume an implementation like the one [here](https://github.com/nventuro/tabookey-gasless/blob/relay-call-no-revert/contracts/RelayHub.sol#L262), where RelayHub reports the error code not via a revert reason but in this same event)

Relay:
 * wants to not waste time (it only sends a few txs per block) or money (on transactions that won't be charged for)
 * will check if the transaction emits `RelayCallFailed`. This happening means it might be being attacked (with e.g. an inconsistent `accept` function, or a recipient that withdrew its balance frontrunning the relay), and it may blacklist the recipient

Client:
 * wants to not waste time (get delayed) on rejected transactions, nor money on failed executions
 * if the transaction emits `RelayCallFailed`, the relay may be doing a denial of service attack, and the client might switch to a different relay (and possibly penalize the original one)
 * if the transaction emits `TransactionRelayed`, but the relayed function reverted, the client:
   - may not care, if the recipient charged the user for this execution, who may have performed an invalid action
    - if it _does_ care, the error originated in the client, and it will be handled in a dapp-specific manner

Note that the client distrusts the relay if `RelayCallFailed` is emitted, but needs to look inward if the relayed call was reverted. The possible reasons behind each failure type are completely different, and should not be bundled together.

In other words, these two events can be thought of as 'did the GSN work or not' (was the transaction relayed). Whatever happened with the relayed transaction itself is always the dapp's concern, and handling of this is separate from handling of GSN-related errors.

File: contracts/IRelayHub.sol (hunk 1 above)
Comment (drortirosh) [inline contracts/IRelayHub.sol]: I think we can do with this events model, just fix the event name: its merely means that `CanRelayFailed`, and should be named so...

File: src/js/relayclient/RelayClient.js [hunk 3]
```diff
@@ -412,8 +412,9 @@ class RelayClient {
     fixTransactionReceiptResp(resp) {
         if (resp && resp.result && resp.result.logs) {
             let logs = abi_decoder.decodeLogs(resp.result.logs);
-            let relayed = logs.find(e => e && e.name == 'TransactionRelayed');
-            if (relayed && relayed.events.find(e => e.name == "status").value != 0) {
+            let relayCallFailed = logs.find(e => e && e.name == 'RelayCallFailed');
+            let transactionRelayed = logs.find(e => e && e.name == 'TransactionRelayed');
+            if (relayCallFailed || (transactionRelayed && transactionRelayed.events.find(e => e.name == "status").value != 0)) {
                 console.log("reverted relayed transaction. changing status to zero");
```
Comment (drortirosh) [inline src/js/relayclient/RelayClient.js]: In case of failure, please report (log) the reason code (from either event..) - as its the only source of that information for a dapp creator (the error also appear in the relay log, but that's accessible only on local relay)

File: src/js/relayclient/RelayClient.js [hunk 4]
```diff
@@ -412,12 +412,19 @@ class RelayClient {
...
-            if (relayCallFailed || (transactionRelayed && transactionRelayed.events.find(e => e.name == "status").value != 0)) {
-                console.log("reverted relayed transaction. changing status to zero");
+
+            const setErrorStatus = (reason) => {
+                console.log(`${reason}. changing status to zero`)
                 resp.result.status = 0
             }
+
+            if (canRelayFailed) {
+                setErrorStatus("canRelay failed")
```
[the remark this answers is not in the document]
Comment (drortirosh) [reply src/js/relayclient/RelayClient.js]: the whole idea of the status code is to get logged, so that a user may understand the failure reason.

[the remark this answers is not in the document]
Comment (nventuro) [reply src/js/relayclient/RelayClient.js]: You're right, this slipped by

**Comment by nventuro** [comment]
I mistakenly deleted the branch on my fork, and it looks like GitHub doesn't allow reopening PRs in such cases. Created https://github.com/tabookey/tabookey-gasless/pull/178
